trieste/acquisition/rule.py

```diff
@@ -1055,10 +1169,38 @@ def acquire(
             points from the previous acquisition state.
         """
 
+        # Subspaces should be set by the time we call `acquire`.
+        assert self._tags is not None
+        assert self._init_subspaces is not None
+
+        num_local_models = Counter(
+            LocalizedTag.from_tag(tag).global_tag
+            for tag in models
+            if LocalizedTag.from_tag(tag).is_local
+        )
+        num_local_models_vals = set(num_local_models.values())
+        assert (
+            len(num_local_models_vals) <= 1
+        ), f"The number of local models should be the same for all tags, got {num_local_models}"
+        _num_local_models = sum(num_local_models_vals)
+
+        num_subspaces = len(self._tags)
+        assert _num_local_models in [0, num_subspaces], (
+            f"When using local models, the number of subspaces {num_subspaces} should be equal to "
+            f"the number of local models {_num_local_models}"
+        )
+
+        # If we have local models, run the (deepcopied) base rule sequentially for each subspace.
```

Comment:
why do we need to copy the rules? is there an issue with iterating through the local models and sending them to the acquire call of a same acquisition rule?

Reply:
We only copy rules when we either have local models, or not using EGO. As the rules compile the acquisition functions, the model trainable variables will be built into the graph, so we need to build separate graphs for local models. Similarly, non-EGO rules may have some state that needs to be maintained between calls for each region, so it is safer to use separate instances.